### PR TITLE
contrib/segmentio/kafka.go.v0: add DSM support

### PR DESCRIPTION
**Pull request by adrien-f**
### What does this PR do?

Add Data Streams Monitoring support to segmentio/kafka integration

### Motivation

We currently use segmentio/kafka for our Go projects and it was easier to update the integration than change library.

### Reviewer's Checklist

- [x] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.

For Datadog employees:

- [ ] If this PR touches code that handles credentials of any kind, such as Datadog API keys, I've requested a review from `@DataDog/security-design-and-guidance`.
- [ ] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: contrib/segmentio/kafka.go.v0/kafka.go
Comment (adrien-f) [inline contrib/segmentio/kafka.go.v0/kafka.go:263]: There is an alternative private implementation here:

https://github.com/segmentio/kafka-go/blob/ebca72eaee918d303c532feb7ff29afdcd8c2efa/message.go#L48-L50

I've kept with the one in the other integrations.

**Comment by darccio** [comment]
Hi @adrien-f, thanks for contributing this feature. I approved running the CI pipeline and there are some things to fix:

- It seems to fail in [`kafka_go.go:210`](https://github.com/DataDog/dd-trace-go/pull/2625/files#diff-d612bf6dbf008c33be8cbcbe58b5ba09af802ba93210ea39141e44a9e26867d1R210) with [this error](https://github.com/DataDog/dd-trace-go/actions/runs/8376564278/job/22937344258?pr=2625#step:5:314).
- A little nitpick [here](https://github.com/DataDog/dd-trace-go/actions/runs/8376564278/job/22937340909?pr=2625#step:3:95): `  contrib/segmentio/kafka.go.v0/kafka.go:61:2: var-naming: struct field groupId should be groupID (revive)`

**Comment by adrien-f** [comment]
Many thanks @darccio !

I've pushed the `groupID` change 👍 

Regarding the issue reported in the CI, it appears to be caused by the `-race` flag, I will investigate what's up with that.

**Comment by adrien-f** [comment]
I believe I corrected the issue 👍 

Happy weekend,

File: contrib/segmentio/kafka.go.v0/kafka.go
Comment (darccio) [inline contrib/segmentio/kafka.go.v0/kafka.go:52]: @adrien-f Comparing with `Shopify/sarama` contrib, `groupID` comes from [`WithGroupID`](https://github.com/DataDog/dd-trace-go/blob/4904f57e55983c8cbf261f9c7b1561fc04071729/contrib/Shopify/sarama/option.go#L63) instead. I think it'll be better to be coherent with the other contrib, and avoid changing the default behaviour (working without GroupID).

Do you mind adding the option and removing this line?

**Review by darccio — CHANGES_REQUESTED** [review]

File: contrib/segmentio/kafka.go.v0/kafka.go
Comment (adrien-f) [reply contrib/segmentio/kafka.go.v0/kafka.go:52]: For sure, I'm also looking at the `confluent-kafka-go` integration and they have a WithConfig: 

https://github.com/DataDog/dd-trace-go/blob/4904f57e55983c8cbf261f9c7b1561fc04071729/contrib/confluentinc/confluent-kafka-go/kafka/option.go#L114-L116

Which is called through this:

https://github.com/DataDog/dd-trace-go/blob/4904f57e55983c8cbf261f9c7b1561fc04071729/contrib/confluentinc/confluent-kafka-go/kafka/kafka.go#L43-L44

And I like this approach of enhancing the constructor transparently as users adopting DSM will only need to update and turn on the environment variable instead of having to explicitly pass the groupID. 

This might be too much of a broad change for this PR though, what do you think?

Comment (darccio) [reply contrib/segmentio/kafka.go.v0/kafka.go:52]: @adrien-f If you want, I can contribute the change. It won't take me too much time. I'd rather avoid introducing a subtle breaking change.

Comment (adrien-f) [reply contrib/segmentio/kafka.go.v0/kafka.go:52]: No worries, I just wanted to be sure :) I've pushed the change!

Comment (darccio) [reply contrib/segmentio/kafka.go.v0/kafka.go:52]: @adrien-f I don't see any new commit.

**Review by darccio — APPROVED** [review]
LGTM

**Comment by adrien-f** [comment]
> LGTM

Thanks for the review :shipit: !

**Comment by darccio** [comment]
> Thanks for the review :shipit: !

You are welcome. I'll get a second review from Ecosystems, @rarguelloF, but it won't be done until next week (Holy week is close, so some Datadogs are already OOO).

File: contrib/segmentio/kafka.go.v0/option.go
Comment (adrien-f) [reply contrib/segmentio/kafka.go.v0/option.go:95]: This is what I had at first before changing it following https://github.com/DataDog/dd-trace-go/pull/2625#discussion_r1537820468

File: contrib/segmentio/kafka.go.v0/kafka.go
Comment (rarguelloF) [reply contrib/segmentio/kafka.go.v0/kafka.go]: @darccio @adrien-f the reason sarama has this group ID option is because the consumer group is not available otherwise (the usage of consumer groups is through a different api that we currently don't support, see: https://github.com/DataDog/dd-trace-go/issues/2133).

In general, the implementation that we should be looking for reference is the confluent-kafka one: https://github.com/DataDog/dd-trace-go/blob/main/contrib/confluentinc/confluent-kafka-go/kafka.v2/option.go#L114-L116 -> the groupID is fetched from the kafka config actually used by the consumer.

File: contrib/segmentio/kafka.go.v0/option.go
Comment (rarguelloF) [reply contrib/segmentio/kafka.go.v0/option.go:95]: I replied here: https://github.com/DataDog/dd-trace-go/pull/2625#discussion_r1547488982

Sorry for the confusion! @adrien-f but TLDR; the reason we have that option in Sarama is because we don't have a way to access the groupID there, but in this contrib and the confluent kafka one we do have the option to access it from the config, so this would make this `WithGroupID` option unnecessary.

**Review by piochelepiotr — APPROVED** [review]

**Review by rarguelloF — APPROVED** [review]

**Comment by darccio** [comment]
@adrien-f Please, merge `main` in your branch. This will trigger the CI and the automerging that I just activated. Thanks!

**Comment by adrien-f** [comment]
@darccio I used the rebase button/feature but looks like it removed the auto-merge, sorry about that 🙏

**Pull request merged**
